pyrealm/core/solar.py

```diff
@@ -1,26 +1,600 @@
+    variables and constants.
+
+    .. math::
+
+        ppfd = (1.0 \times 10^{-6}) \cdot k\_fFEC \cdot (1.0 - k\_alb\_vis) \cdot \tau \
+            cdot ra\_d
+
+    Args:
+        tau          : bulk transmissivity, unitless
+        ra_d         : daily solar radiation, J/m^2
+        k_fFEC       : from flux to energy conversion, umol/J
+        k_alb_vis    : visible light albedo
+
+    Returns:
+        An array of photosynthetic photon flux density, mol/m^2
+    """
+
+    ppfd = (1.0e-6) * k_fFEC * (1.0 - k_alb_vis) * tau * ra_d
+
+    return ppfd
+
+
+def calc_ppfd(
+    sf: NDArray,
+    elv: NDArray,
+    latitude: NDArray,
+    julian_day: NDArray,
+    n_days: NDArray,
+    const: CoreConst,
+) -> NDArray:
+    """Calculates photosynthetic photon flux density, PPFD,(mol/m^2).
+
+    This function calulates ppfd (mol/m^2) from primary variables.
+
+    Args:
+        sf           : Daily sunshine fraction of observations, unitless
+        elv          : Elevation of observations, metres
+        latitude     : The Latitude of observations (degrees)
+        julian_day   : Julian day of the year
+        n_days       : days in the calendar year
+        const        : CoreConst object containing core constants:
+
+    Returns:
+        An array of photosynthetic photon flux density, PPFD,(mol/m^2)
+
+    Example:
+        >>> # Calculate ppfd
+        >>> import numpy as np
+        >>> from pyrealm.constants import CoreConst
+        >>> # Create dataclass of constants
+        >>> const = CoreConst()
+        >>> # define variable values
+        >>> sf = np.array([1.0])
+        >>> elv = np.array([142])
+        >>> latitude = np.array([37.7])
+        >>> julian_day = np.array([172])
+        >>> n_days = np.array([366])
+        >>> # evaluate function
+        >>> calc_ppfd(sf = sf, elv = elv, latitude = latitude, julian_day = julian_day,\
+            n_days = n_days, const = const)
+        array([62.04230021])
+
+
+    """
+    # Validate the inputs
+    _ = check_input_shapes(latitude, elv, sf, julian_day, n_days)
+
+    # Calculate intermediate values
+
+    nu, lambda_ = calc_heliocentric_longitudes(julian_day=julian_day, n_days=n_days)
+
+    dr = calc_distance_factor(nu=nu, k_e=const.k_e)
+
+    delta = calc_declination_angle_delta(
+        lambda_=lambda_, k_eps=const.k_eps, k_pir=const.k_pir
+    )
+
+    tau = calc_transmissivity(sf=sf, elv=elv, k_c=const.k_c, k_d=const.k_d)
+
+    hs = calc_sunset_hour_angle(delta=delta, latitude=latitude, k_pir=const.k_pir)
+
+    ra_d = calc_daily_solar_radiation(
+        dr=dr, hs=hs, delta=delta, latitude=latitude, const=const
+    )
+
+    # Calculate ppfd
+
+    ppfd = calc_ppfd_from_tau_ra_d(
+        tau=tau, ra_d=ra_d, k_fFEC=const.k_fFEC, k_alb_vis=const.k_alb_vis
+    )
+
+    return ppfd
+
+
+def calc_rnl(sf: NDArray, tc: NDArray, k_b: float, k_A: float) -> NDArray:
+    r"""Calculates net longwave radiation, rnl, W/m^2.
```

Comment:
I've a preference for more expressive function names:  `calc_net_longwave_radiation`

Reply:
Sorted